src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java

```diff
@@ -319,191 +316,91 @@ static void verifyColumnMasterKeyMetadata(SQLServerConnection connection, SQLSer
         }
     }
 
-    private static SimpleTtlCache<String, SqlFedAuthToken> msiTokenCache = new SimpleTtlCache<>();
-
     /**
-     * Get Managed Identity Authentication token
+     * Get Managed Identity Authentication token through a ManagedIdentityCredential
      * 
      * @param resource
      *        token resource
      * @param msiClientId
      *        Managed Identity or User Assigned Managed Identity
-     * @param tokenCacheTtl
-     *        The number of seconds the token should remain in the cache
      * @return fedauth token
      * @throws SQLServerException
      */
-    static SqlFedAuthToken getMSIAuthToken(String resource, String msiClientId,
-            int tokenCacheTtl) throws SQLServerException {
-        String cacheKey = "resource:" + resource + "|clientid:" + msiClientId;
-        SqlFedAuthToken token = msiTokenCache.get(cacheKey);
-        if (token != null) {
-            if (connectionlogger.isLoggable(Level.FINER)) {
-                connectionlogger.finer("Using cached Managed Identity auth token: " + token.toString());
-            }
-            return token;
+    static SqlFedAuthToken getManagedIdentityCredAuthToken(String resource, String msiClientId) {
+        ManagedIdentityCredential mic = null;
+
+        if (null != msiClientId && !msiClientId.isEmpty()) {
+            mic = new ManagedIdentityCredentialBuilder().clientId(msiClientId).build();
+        } else {
+            mic = new ManagedIdentityCredentialBuilder().build();
         }
 
-        // IMDS upgrade time can take up to 70s
-        final int imdsUpgradeTimeInMs = 70 * 1000;
-        final List<Integer> retrySlots = new ArrayList<>();
+        TokenRequestContext tokenRequestContext = new TokenRequestContext();
+        String scope = resource.endsWith(SQLServerMSAL4JUtils.SLASH_DEFAULT) ? resource : resource + SQLServerMSAL4JUtils.SLASH_DEFAULT;
+        tokenRequestContext.setScopes(Arrays.asList(scope));
 
-        StringBuilder urlString = new StringBuilder();
-        int retry = 1, maxRetry = 1;
+        SqlFedAuthToken sqlFedAuthToken = null;
 
-        // MSI_ENDPOINT and MSI_SECRET can be used instead of IDENTITY_ENDPOINT and IDENTITY_HEADER
-        String identityEndpoint = System.getenv("IDENTITY_ENDPOINT");
-        if (null == identityEndpoint || identityEndpoint.trim().isEmpty()) {
-            identityEndpoint = System.getenv("MSI_ENDPOINT");
-        }
+        Optional<AccessToken> accessTokenOptional = mic
+                .getToken(tokenRequestContext)
+                .blockOptional();
 
-        String identityHeader = System.getenv("IDENTITY_HEADER");
-        if (null == identityHeader || identityHeader.trim().isEmpty()) {
-            identityHeader = System.getenv("MSI_SECRET");
+        if (!accessTokenOptional.isPresent()) {
+            if (connectionlogger.isLoggable(java.util.logging.Level.FINE)) {

```

Comment:
If there is no result, no error, and no token, we'll need to throw our own exception. We can either throw it here or in the calling method. The calling method (or this one) should never return a null result.